// material-ui-fullscreen-dialog/lib/FullscreenDialogFrame.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = FullscreenDialogFrame;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _TransitionGroup = require('react-transition-group/TransitionGroup');

var _TransitionGroup2 = _interopRequireDefault(_TransitionGroup);

var _AutoLockScrolling = require('material-ui/internal/AutoLockScrolling');

var _AutoLockScrolling2 = _interopRequireDefault(_AutoLockScrolling);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {
  root: {
    width: '100vw',
    height: '100vh',
    position: 'fixed',
    top: 0,
    left: 0,
    zIndex: 1500,
    background: '#fafafa'
  }
};

var TransitionItem = function (_Component) {
  _inherits(TransitionItem, _Component);

  function TransitionItem(props) {
    _classCallCheck(this, TransitionItem);

    var _this = _possibleConstructorReturn(this, (TransitionItem.__proto__ || Object.getPrototypeOf(TransitionItem)).call(this, props));

    _this.state = {
      style: {
        opacity: 0,
        transition: 'all 225ms cubic-bezier(0.0, 0.0, 0.2, 1)',
        transform: 'translate(0, 56px)'
      }
    };
    return _this;
  }

  _createClass(TransitionItem, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      clearTimeout(this.enterTimeout);
      clearTimeout(this.leaveTimeout);
    }
  }, {
    key: 'componentWillEnter',
    value: function componentWillEnter(callback) {
      this.componentWillAppear(callback);
    }
  }, {
    key: 'componentWillAppear',
    value: function componentWillAppear(callback) {
      var _this2 = this;

      this.enterTimeout = setTimeout(function () {
        _this2.setState({
          style: {
            opacity: 1,
            transition: 'all 225ms cubic-bezier(0.0, 0.0, 0.2, 1)',
            transform: 'translate(0, 0px)'
          }
        });
        _this2.enterTimeout = setTimeout(callback, 225 // matches transition duration
        );
      });
    }
  }, {
    key: 'componentWillLeave',
    value: function componentWillLeave(callback) {
      this.setState({
        style: {
          opacity: 0,
          transition: 'all 195ms cubic-bezier(0.4, 0.0, 1, 1)',
          transform: 'translate(0, 56px)'
        }
      });

      this.leaveTimeout = setTimeout(callback, 195 // matches transition duration
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          style = _props.style,
          children = _props.children,
          other = _objectWithoutProperties(_props, ['style', 'children']);

      return _react2.default.createElement(
        'div',
        _extends({}, other, { style: _extends({}, style, this.state.style) }),
        children
      );
    }
  }]);

  return TransitionItem;
}(_react.Component);

TransitionItem.propTypes = {
  children: _propTypes2.default.node,
  style: _propTypes2.default.object
};

function FullscreenDialogFrame(_ref) {
  var children = _ref.children,
      open = _ref.open,
      style = _ref.style;

  return _react2.default.createElement(
    _TransitionGroup2.default,
    {
      component: 'div',
      transitionAppear: true,
      transitionAppearTimeout: 225,
      transitionEnter: true,
      transitionEnterTimeout: 225
    },
    open && _react2.default.createElement(
      TransitionItem,
      { style: _extends({}, styles.root, style)
      },
      children
    ),
    _react2.default.createElement(_AutoLockScrolling2.default, { lock: open })
  );
}

FullscreenDialogFrame.propTypes = {
  children: _propTypes2.default.node,
  open: _propTypes2.default.bool.isRequired,
  style: _propTypes2.default.object
};
